src/tokens/TokenImport.tsx

```diff
@@ -27,58 +30,175 @@ import variables from 'src/styles/variables'
+        networkId,
+        tokenId,
+        tokenAddress,
+        error: Errors.AlreadySupported,
+      })
+      return false
     }
-  }
 
-  const handleAddressFocus = () => {
-    setIsCompleteAddress(false)
     setError(null)
+    return true
   }
 
+  const fetchTokenDetails = async (): Promise<TokenDetails> => {
+    if (!isAddress(tokenAddress)) {
+      // shouldn't happen as this function is called only after validating the address
+      throw new Error('Invalid token address')
+    }
+    if (!walletAddress || !isAddress(walletAddress)) {
+      // should never happen
+      throw new Error('No wallet address found when fetching token details')
+    }
+
+    const client = publicClient[networkIdToNetwork[networkId]]
+    const contract = getContract({
+      abi: erc20.abi,
+      address: tokenAddress,
+      publicClient: client,
+    })
+
+    const contractCode = await client.getBytecode({ address: tokenAddress })
+    if (!contractCode) throw new NotContractError('Contract code is empty')
+
+    const [symbol, decimals, name, balance] = await Promise.all([
+      contract.read.symbol(),
+      contract.read.decimals(),
+      contract.read.name(),
+      contract.read.balanceOf([walletAddress]),
+    ])
+    return { address: tokenAddress, symbol, decimals, name, balance }
+  }
+
+  const validateContract = useAsyncCallback(fetchTokenDetails, {
+    onSuccess: (details) => {
+      Logger.info(
+        TAG,
+        `Wallet ${walletAddress} holds ${formatUnits(details.balance, details.decimals)} ${
+          details.symbol
+        } (${details.name} = ${details.address})})`
+      )
+      setTokenSymbol(details.symbol)
+    },
+    onError: (error) => {
+      Logger.error(TAG, `Could not fetch token details`, error)
+      const tokenId = getTokenId(networkId, tokenAddress.toLowerCase())
+
+      // it doesn't directly show if it's a timeout error, need to check cause recursively
+      const hasTimeout = (error: unknown): boolean =>
+        error instanceof BaseError &&
+        (error instanceof TimeoutError || (error.cause !== undefined && hasTimeout(error.cause)))
+      if (hasTimeout(error)) {
+        setError(t('tokenImport.error.invalidToken'))
+        ValoraAnalytics.track(AssetsEvents.import_token_error, {
+          networkId,
+          tokenId,
+          tokenAddress,
+          error: Errors.Timeout,
+        })
+        return
+      }
+
+      if (error instanceof NotContractError) {
+        setError(t('tokenImport.error.invalidToken'))
+        ValoraAnalytics.track(AssetsEvents.import_token_error, {
+          networkId,
+          tokenId,
+          tokenAddress,
+          error: Errors.NotContract,
+        })
+        return
+      }
+
+      setError(t('tokenImport.error.invalidToken'))
+      ValoraAnalytics.track(AssetsEvents.import_token_error, {
+        networkId,
+        tokenId,
+        tokenAddress,
+        error: Errors.NotERC20,
+      })
+    },
```

Comment:
nit: this error block could be simplified by having something (function / ternary / code block) that decides on the type of error, and then set the error + dispatch the analytics with the error type? it's easier to quickly read the intention of the function that way, currently it seems like depending on the type of error we may be doing custom actions

Reply:
Similar feedback to https://github.com/valora-inc/wallet/pull/4618#discussion_r1424216258

Updated now! It was a remnant of those other error messages and I didn't refactor this code.